discord/bot.py

```diff
@@ -0,0 +1,184 @@
+from .shard import AutoShardedClient
+from .utils import get
+from .app import SlashCommand, SubCommandGroup, MessageCommand, UserCommand, ApplicationCommand
+
+class ApplicationCommandMixin:
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.to_register = []
+        self.app_commands = {}
+
+    def add_application_command(self, command):
+        """Adds a :class:`.ApplicationCommand` into the internal list of commands.
+
+        This is usually not called, instead the :meth:`~.ApplicationMixin.command` or
+        other shortcut decorators are used instead.
+
+        Parameters
+        -----------
+        command: :class:`.ApplicationCommand`
+            The command to add.
+        """
+        self.to_register.append(command)
+
+    def remove_application_command(self, command):
+        self.app_commands.remove(command)
+
+    async def sync_commands(self):
+        to_add = [i for i in self.to_register] + [i for i in self.app_commands.values()]
+        cmds = await self.http.bulk_upsert_global_commands(
+            self.user.id,
+            [i.to_dict() for i in self.to_register]
+            + [i.to_dict() for i in self.app_commands.values()],
+        )
+        new_cmds = {}
+        self.app_commands = {}
+        for i in cmds:
+            cmd = get(to_add, name=i["name"], description=i["description"], type=i["type"])
+            new_cmds[i["id"]] = cmd
+
+    async def register_commands(self):
+        """|coro|
+        Needs documentation
+
+        By default, this coroutine is called inside the :func:`.on_connect`
+        event. If you choose to override the :func:`.on_connect` event, then
+        you should invoke this coroutine as well.
+        """
+        commands = []
+
+        registered_commands = await self.http.get_global_commands(self.user.id)
+        for command in [cmd for cmd in self.to_register if cmd.guild_ids is None]:
+            as_dict = command.to_dict()
+            if not len(registered_commands) == 0:
+                matches = [x for x in registered_commands if x["name"] == command.name and x['type'] == command.type]
+                # TODO: rewrite this, it seems inefficient
+                if len(matches) > 0:
+                    as_dict['id'] = matches[0]["id"]
+            commands.append(as_dict)
+
+        update_guild_commands = {}
+        async for guild in self.fetch_guilds(limit=None):
+            update_guild_commands[guild.id] = []
+        for command in [cmd for cmd in self.to_register if cmd.guild_ids is not None]:
+            as_dict = command.to_dict()
+            for guild_id in command.guild_ids:
+                to_update = update_guild_commands[guild_id]
+                update_guild_commands[guild_id] = to_update + [as_dict]
+
+        for guild_id in update_guild_commands:
+            cmds = await self.http.bulk_upsert_guild_commands(self.user.id, guild_id, update_guild_commands[guild_id])
```

Comment:
This should only be called if `update_guild_commands[guild_id]` is not empty, no?